src/cargo/core/resolver/mod.rs

```diff
@@ -1098,3 +1099,28 @@ fn check_cycles(resolve: &Resolve, activations: &Activations) -> CargoResult<()>
         Ok(())
     }
 }
+
+fn get_duplicate_pkgs(resolve: &Resolve) -> Vec<&'static str> {
+    let mut unique_pkg_ids = HashSet::new();
+    let mut result = HashSet::new();
+    for pkg_id in resolve.iter() {
+        let mut encodable_pkd_id = encode::encodable_package_id(pkg_id);
```

Comment:
Let's add a comment here, explaining why we need to encode `package_id`s? Something like this:

```
When writing package id's to lockfile, we apply lossy encoding. In particular, we don't store paths of path dependencies.  That means that *different* package may collide in the lockfile, hence this check.
```

Reply:
Refactored the function and added the comment as doc string.